Document/0x05e-Testing-Cryptography.md

```diff
@@ -291,6 +291,74 @@ When defining the KeyDescription AuthorizationList, the following parameters wil
+
+During key attestation, we can specify the alias of a key pair and in return, get a certificate chain, which we can use to verify the properties of that key pair. If the root certificate of the chain is [Google Hardware Attestation Root certificate]("https://developer.android.com/training/articles/security-key-attestation#root_certificate" "Google Hardware Attestation Root certificate") and the checks related to key pair storage in hardware are made it gives an assurance that the device supports hardware-level key attestation and the key is in hardware-backed keystore that Google believes to be secure. Alternatively, if the attestation chain has any other root certificate, then Google does not make any claims about the security of the hardware.
+
+
+Although the key attestation process can be implemented within the application directly but it is recommended that it should be implemented at the server-side for security reasons. The following are the high-level guidelines for the secure implementation of Key Attestation:
+
+* The server should initiate the key attestation process by creating a random number securely using CSPRNG(Cryptographically Secure Random Number Generator) and the same should be sent to the user as a challenge. 
+
+* The client should call the _setAttestationChallenge()_ API with the challenge received from the server and should then retrieve the attestation certificate chain using the _KeyStore.getCertificateChain()_ method.
+
+* The attestation response should be sent to the server for the verification and following checks should be performed for the verification of the key attestation response:
+
+  * Verify the certificate chain, up to the root and perform certificate sanity checks such as validity, integrity and trustworthiness.
+
+  * Check if the root certificate is signed with the Google attestation root key which makes the attestation process trustworthy.
+
+  * Extract the attestation certificate extension data, which appears within the first element of the certificate chain and perform the following checks:
+
+    * Verify that the attestation challenge is having the same value which was generated at the server while initiating the attestation process.
+
+    * Verify the signature in the key attestation response.
+
+    * Now check the Keymaster(Keymaster is a software that runs in the security context and provides all the secure keystore operations) security level which will be one of Software, TrustedEnvironment or StrongBox to determine if the device has secure key storage mechanism.
+
+    * Additionally, you can check the attestation security level which will be one of Software, TrustedEnvironment or StrongBox to check how the attestation certificate was generated. Also, some other checks pertaining to keys can be made such as purpose, access time, authentication requirement, etc. to verify the key attributes.
+
+The typical example of Android Keystore attestation looks like this:
```

Comment:
The typical example of Android Keystore attestation looks like this: -->
he typical example of Android Keystore attestation response looks like this: